src/module_4/module_4_model-fitting.md

```diff
@@ -0,0 +1,1214 @@
+      <td>2020-10-05 00:00:00</td>
+      <td>2</td>
+      <td>0.0</td>
+      <td>0.0</td>
+      <td>0.0</td>
+      <td>...</td>
+      <td>0.0</td>
+      <td>0.0</td>
+      <td>0.0</td>
+      <td>2.0</td>
+      <td>33.0</td>
+      <td>42.0</td>
+      <td>31.134053</td>
+      <td>30.0</td>
+      <td>30.0</td>
+      <td>24.27618</td>
+    </tr>
+    <tr>
+      <th>2</th>
+      <td>33826472919172</td>
+      <td>ricepastapulses</td>
+      <td>2808099078276</td>
+      <td>3481384026244</td>
+      <td>2020-10-05 20:08:53</td>
+      <td>2020-10-05 00:00:00</td>
+      <td>4</td>
+      <td>0.0</td>
+      <td>0.0</td>
+      <td>0.0</td>
+      <td>...</td>
+      <td>0.0</td>
+      <td>0.0</td>
+      <td>0.0</td>
+      <td>2.0</td>
+      <td>33.0</td>
+      <td>42.0</td>
+      <td>31.134053</td>
+      <td>30.0</td>
+      <td>30.0</td>
+      <td>24.27618</td>
+    </tr>
+    <tr>
+      <th>3</th>
+      <td>33826472919172</td>
+      <td>ricepastapulses</td>
+      <td>2808393957508</td>
+      <td>3291363377284</td>
+      <td>2020-10-06 08:57:59</td>
+      <td>2020-10-06 00:00:00</td>
+      <td>2</td>
+      <td>0.0</td>
+      <td>0.0</td>
+      <td>0.0</td>
+      <td>...</td>
+      <td>0.0</td>
+      <td>0.0</td>
+      <td>0.0</td>
+      <td>2.0</td>
+      <td>33.0</td>
+      <td>42.0</td>
+      <td>31.134053</td>
+      <td>30.0</td>
+      <td>30.0</td>
+      <td>24.27618</td>
+    </tr>
+    <tr>
+      <th>4</th>
+      <td>33826472919172</td>
+      <td>ricepastapulses</td>
+      <td>2808429314180</td>
+      <td>3537167515780</td>
+      <td>2020-10-06 10:37:05</td>
+      <td>2020-10-06 00:00:00</td>
+      <td>3</td>
+      <td>0.0</td>
+      <td>0.0</td>
+      <td>0.0</td>
+      <td>...</td>
+      <td>0.0</td>
+      <td>0.0</td>
+      <td>0.0</td>
+      <td>2.0</td>
+      <td>33.0</td>
+      <td>42.0</td>
+      <td>31.134053</td>
+      <td>30.0</td>
+      <td>30.0</td>
+      <td>24.27618</td>
+    </tr>
+  </tbody>
+</table>
+<p>5 rows × 27 columns</p>
+</div>
+
+
+
+
+```python
+min_products = 5
+order_size = df_feature_frame.groupby('order_id')['outcome'].sum()
+orders_of_min_size = order_size[order_size >= min_products].index
+df_filtered = df_feature_frame[df_feature_frame['order_id'].isin(orders_of_min_size)]
+```
+
+
+```python
+info_cols = ['variant_id', 'order_id', 'user_id', 'created_at', 'order_date']
+label_col = 'outcome'
+features_cols = [col for col in df_filtered.columns if col not in info_cols + [label_col]]
+
+categorical_cols = ['product_type', 'vendor']
+binary_cols = ['ordered_before', 'abandoned_before', 'active_snoozed', 'set_as_regular']
+numerical_cols = [col for col in features_cols if col not in categorical_cols + binary_cols]
+```
+
+
+```python
+df_filtered = df_filtered.sort_values('created_at')
+
+order_ids = df_filtered['order_id'].unique()
+
+total_ids = len(order_ids)
+train_end = int(total_ids * 0.7)
+val_end = train_end + int(total_ids * 0.15)
```

Comment:
This evaluation schema is in the good direction but I would consider:

- make the train/eval/test split over days instead of over orders (bear in mind that in the train period you are NOT supposed to know how many orders you are gonna have in the future, while with the schema that you are proposing you are implicitly assuming this number, this is not true if you just consider number of days)
- Have a rolling window where you keep moving the train / test to the right and you train and evaluate in different months to see how the model behaves in different periods